src/btree/bt_vrfy.c

```diff
@@ -216,37 +216,59 @@ __wt_verify(WT_SESSION_IMPL *session, const char *cfg[])
 		    ckpt->raw.data, ckpt->raw.size,
 		    root_addr, &root_addr_size, true));
 
-		/*
-		 * Ignore trees with no root page.
-		 * Verify, then discard the checkpoint from the cache.
-		 */
-		if (root_addr_size != 0 &&
-		    (ret = __wt_btree_tree_open(
-		    session, root_addr, root_addr_size)) == 0) {
+		/* Skip trees with no root page. */
+		if (root_addr_size != 0) {
+			WT_ERR(__wt_btree_tree_open(
+			    session, root_addr, root_addr_size));
+
 			if (WT_VRFY_DUMP(vs))
 				WT_ERR(__wt_msg(session, "Root: %s %s",
 				    __wt_addr_string(session,
 				    root_addr, root_addr_size, vs->tmp1),
 				    __wt_page_type_string(
 				    btree->root.page->type)));
 
+			__wt_evict_file_exclusive_off(session);
+
+			/* Verify the tree. */
 			WT_WITH_PAGE_INDEX(session,
 			    ret = __verify_tree(session, &btree->root, vs));
 
+			/*
+			 * We have an exclusive lock on the handle, but we're
+			 * swapping root pages in-and-out of that handle, and
+			 * there's a race with eviction entering the tree and
+			 * seeing an invalid root page. Eviction must work on
+			 * trees being verified (else we'd have to do our own
+			 * eviction), lock eviction out whenever we're loading
+			 * a new root page. This loops works because we are
+			 * called with eviction locked out, so we release the
+			 * lock at the top of the loop and re-acquire it here.
+			 */
+			WT_TRET(__wt_evict_file_exclusive_on(session));
 			WT_TRET(__wt_cache_op(session, WT_SYNC_DISCARD));
 		}
 
 		/* Unload the checkpoint. */
 		WT_TRET(bm->checkpoint_unload(bm, session));
+
+		/*
+		 * We've finished one checkpoint's verification (verification,
+		 * then cache eviction and checkpoint unload): if any errors
+		 * occurred, quit. Done this way because otherwise we'd need
+		 * at least two more state variables on error, one to know if
+		 * we need to discard the tree from the cache and one to know
+		 * if we need to unload the checkpoint.
+		 */
 		WT_ERR(ret);
 
 		/* Display the tree shape. */
 		if (vs->dump_layout)
 			WT_ERR(__verify_layout(session, vs));
 	}
 
-done:
-err:	/* Inform the underlying block manager we're done. */
```

Comment:
Just curious, why reorder?  I'd normally go for alpha order in the absence of some compelling reason...